neps/nep-0366.md

```diff
@@ -0,0 +1,148 @@
+Author: Illia Polosukhin <ilblackdragon@gmail.com>
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/366
+Status: Draft
+Type: Protocol Track
+Category: Runtime
+Created: 21-Jun-2022
+---
+
+## Summary
+
+In-protocol meta transactions allowing for third-party account to initiate and pay transaction fees on behalf of the account.
+
+## Motivation
+
+NEAR has been designed with simplicity of onboarding in mind. One of the large hurdles right now is that after creating an implicit or even named account the user does not have NEAR to pay gas fees to interact with apps.
+
+For example, apps that pay user for doing work (like NEARCrowd or Sweatcoin) or free-to-play games.
+
+[Aurora Plus](https://aurora.plus) has shown viability of the relayers that can offer some number of free transactions and a subscription model. Shifting the complexity of dealing with fees to the infrastructure from the user space.
+
+## Rationale and alternatives
+
+Proposed here design provides the easiest for users and developers way to onboard and pay for user transactions.
+There is no requirement on the user account, including user account may not even exist on chain and implicit account can be used.
+
+An alternative is proxy contracts deployed on the user account.
+This design has severe limitations as it requires user to deploy such contract and additional costs for storage.
+
+## Specification
+
+The main flow of the meta transaction will be as follows:
+ - User will sign `DelegateActionMessage` specifying set of actions that they need to be executed. It also includes specific relayer to ensure secure execution.
+ - User sends signed `DelegateAction` data to the relayer
+ - Relayer forms a `Transaction` with `receiver_id` equal to the user's account and `actions: [DelegateAction { ... }]`, and signs it with it's key. Note, that such transaction can contain other actions toward user's account (for example calling a function).
+ - This transaction is processed normally, by creating a receipt with copy of the all the actions into the `Receipt`.
+ - When processing `DelegateAction` a number of checks are done (see below), mainly `signature` matching user account's key.
+ - When such `Receipt` with valid `DelegateAction` in actions arrives to the user's account it gets executed. The executed means creation of a new Receipt with `receiver_id: AccountId`, `actions: Action` matching `receiver_id` and `actions` inside `DelegateAction`. 
+ - The new `Receipt` looks like normal receipt that would have been originating from user's account, with `predeccessor_id` equal to user's account.
+
+### DelegateAction
+
+Delegate action allows for an account to initiate a batch of actions on behalf of the receiving account, allowing to proxy actions. This is used in implementation of meta transactions.
+
+```rust
+pub struct DelegateAction {
+    /// Receiver of the delegated actions.
+    receiver_id: AccountId,
+    /// List of actions to be executed.
+    actions: Vec<Action>,
```

Comment:
`Vec<Action>` here leads to a type recursion:
```
pub struct Transaction {
   ....
   pub actions: Vec<Action>, // <--- Recursion
}

pub enum Action {
   ...
   Delegate(DelegateAction),
}

pub struct DelegateAction {
   ...
   pub actions: Vec<Action>, /// <--- Recursion
}
```

Reply:
It is acceptable in this context because an Action can't refer to itself or an ancestor, so recursion is finite.

Reply:
I meant, rust compiler would not compile such code because of the recursion.
Example code in playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a6c54d6ed7588eb03d2ad19da1c0a9c6

Reply:
Indeed, a common way to prevent this recursion is to use a `Box`.  As an example, a singly-linked list in Rust can be represented as:
```
struct Node<T> {
    element: T,
    next: Option<Box<Node<T>>>,
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=14f24298eee755a2a08c7b5b66d8b4fa suggests a fix.

Reply:
You are right. When I wrote my comment, I didn't know that Borsh supported `Box`. But there is another issue with this structure: Borsh doesn't support a recursive types https://github.com/near/borsh-rs/pull/96